api/v1alpha1/progressiverollout_types.go

```diff
@@ -28,18 +30,68 @@ type ProgressiveRolloutSpec struct {
 	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
 
-	// Foo is an example field of ProgressiveRollout. Edit ProgressiveRollout_types.go to remove/update
-	Foo string `json:"foo,omitempty"`
+	// SourceRef references the resource, example an ApplicationSet, which owns ArgoCD Applications
+	//+kubebuilder:validation:Required
+	SourceRef corev1.TypedLocalObjectReference `json:"sourceRef"`
+	// Stages reference a list of Progressive Rollout stages
+	//+kubebuilder:validation:Optional
+	Stages []*ProgressiveRolloutStage `json:"stages,omitempty"`
+}
+
+// ProgressiveRolloutStage defines a rollout stage
+type ProgressiveRolloutStage struct {
+	// Name is a human friendly name for the stage
+	//+kubebuilder:validation:Required
+	Name string `json:"name"`
+	// MaxParallel is how many selected targets to update in parallel
+	//+kubebuilder:validation:Minimum:1
+	MaxParallel intstr.IntOrString `json:"maxParallel"`
+	// MaxTargets is the maximum number of selected targets to update
+	//+kubebuilder:validation:Minimum:1
+	MaxTargets intstr.IntOrString `json:"maxTargets"`
+	// Targets is the targets to update in the stage
+	//+kubebuilder:validation:Optional
+	Targets ProgressiveRolloutTargets `json:"targets,omitempty"`
+}
+
+// ProgressiveRolloutTargets defines the target of the Progressive Rollout
+type ProgressiveRolloutTargets struct {
+	// Clusters is the a cluster type of targets
+	//+kubebuilder:validation:Optional
+	Clusters Cluster `json:"clusters"`
+}
+
+// Cluster defines a target of type cluster
+type Cluster struct {
+	// Selector is a label selector to get the clusters for the update
+	//+kubebuilder:validation:Required
+	Selector metav1.LabelSelector `json:"selector"`
 }
 
 // ProgressiveRolloutStatus defines the observed state of ProgressiveRollout
 type ProgressiveRolloutStatus struct {
 	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
 }
 
-// +kubebuilder:object:root=true
+// GetStatusConditions returns a pointer to the Status.Conditions slice
+func (in *ProgressiveRollout) GetStatusConditions() *[]metav1.Condition {
+	return &in.Status.Conditions
+}
 
+func (in *ProgressiveRollout) NewStatusCondition(t string, s metav1.ConditionStatus, r string, m string) metav1.Condition {
```

Comment:
Should the type be its own type and we could use constant to keep consistency across conditions?
(that is, if i understand it well, it would be something like "Deploying", "Deployed", "Failed" or "RolledBack"?

Reply:
The reason why I'm using `t string` (t as type) it's because `metav1.Condition` takes a string as type.

My understanding is that if we were using our own type, we would have a type mismatch, unless there is something I am missing.

Reply:
Gotcha. we'll keep it simple on the conditions